(#1947) Improve support for MongoDB authentication

fatmcgav opened the pull request:
Improve support for MongoDB authentication by:
* Creating an 'administration' user. 
* Store those credentials in '.mongorc.js', which is used by Puppet to authenticate
* Add some tests for this functionality.

fatmcgav commented:
Looks like the test failures are lint issues that are un-related to my changes :(

tphoney commented:
Thanks for the effort could you please squash your commits, it makes the history a lot nicer to look at.

fatmcgav commented:
@tphoney Have just squashed the commits down. 

Let me know if you need anything further...

tphoney commented:
unfortunately this cannot be merged until the travis passes.

fatmcgav commented:
@tphoney Ah, apologies, I was mis-reading the travis failure... Will get it fixed now :)

fatmcgav commented:
@tphoney The latest failures with future_parser don't appear to be related to any of the changes I've made... 

So not sure if this can be merged in as is... LMK :)

fatmcgav commented:
@tphoney Did you need anything more from me on this one? 

Cheers
Gav

tphoney commented:
I really cannot merge this until travis CI passes.

fatmcgav commented:
@tphoney Understood :) However I don't think the changes I've made would affect the travis CI outcome, which coincidently seems to have disappeared... 

I'll get Travis set-up on my fork and see if I can get stuff fixed up... 

Cheers
Gav

fatmcgav commented:
@tphoney I'm just taking a look at this now, and it looks like by merging PR #190 I get some way towards green tests... However some of the FUTURE_PARSER tests are failing with:
> Evaluation Error: Missing title. The title expression resulted in undef

as shown here: https://travis-ci.org/fatmcgav/puppetlabs-mongodb/jobs/60506360#L377

Not sure where to start on that one? 

Are you on IRC so we can chat over? Ping me on same handle in #puppet-dev or #puppet-users... 

Cheers
Gav

fatmcgav commented:
Hmm... Have tracked the issue to this block in server/config.pp: https://github.com/fatmcgav/puppetlabs-mongodb/blob/support_auth/manifests/server/config.pp#L128-L139

However I've tracked `$rcfile` and `$store_creds` through the various classes and can't see anything obvious wrong with it...

fatmcgav commented:
Done some more digging, and it looks like it's not evaluating the `inherits mongodb::params` in server.pp (https://github.com/fatmcgav/puppetlabs-mongodb/blob/support_auth/manifests/server.pp#L77)

As if I explicitly set the `rcfile` value in the pre_condition, the test passes... 

Any ideas? 

cc @hunner / @cmurphy 

Cheers
Gavin

kakalos12 commented:
Hi, 
Until now, is there any way that I can create the admin user ? I always get the error : 

Error: Could not prefetch mongodb_database provider 'mongodb': Execution of '/usr/bin/mongo admin --quiet --port  --eval printjson(db.getMongo().getDBs())' returned 252: 2015-04-30T01:15:04.160+0700 E QUERY    Error: listDatabases failed:{
	"ok" : 0,
	"errmsg" : "not authorized on admin to execute command { listDatabases: 1.0 }",
	"code" : 13
}
    at Error (<anonymous>)
    at Mongo.getDBs (src/mongo/shell/mongo.js:47:15)
    at (shell eval):1:25 at src/mongo/shell/mongo.js:47

fatmcgav commented:
@kakalos12 No, that's what this PR is trying to achieve :)

fatmcgav commented:
@tphoney Right, I've managed to get the tests passing on another branch by merging in PR #190 and making a couple of tweaks to Gemfile and the tests... 

Take a look at: https://github.com/fatmcgav/puppetlabs-mongodb/compare/support_auth...PR190?expand=1

If you're happy, I'll get the tweaks applied to my support_auth branch and rebased... 

Would be good to understand why the `inherits mongodb::params` isn't being honoured though... 

Gav

fatmcgav commented:
@hunner Cheers for merging #210.

I'll rebase this tomorrow and hopefully it'll go green...

fatmcgav commented:
@tphoney/@hunner It's GREEEEENNNNNNNNNNN!!!

Lmk if need anything more from me before merging :)

zeysh commented on manifests/server.pp:
```diff
@@ -89,4 +94,19 @@
+      username      => $admin_username,
+      password_hash => mongodb_password($admin_username, $admin_password),
+      database      => 'admin',
+      roles         => ['dbAdmin', 'dbOwner', 'userAdmin', 'userAdminAnyDatabase'],
```
Hello, 
It seems that 'root' role is need to pass commands for replica config:
http://docs.mongodb.org/manual/reference/built-in-roles/#root

zeysh commented:
Hi @fatmcgav and many thanks !

I have made some tests with your PL and this [doc] (http://docs.mongodb.org/manual/tutorial/deploy-replica-set-with-auth/).

In my vagrant environment, I build 3 vms: vm1 (primary), vm2 (secondary), vm3 (arbiter).
Because vm1 initialise the replicaset, I first create vm3, vm2 and the last one vm1.

db1 (conf sample)
```yaml
[...]
mongodb::server::create_admin: 'true'
mongodb::server::admin_username: admin
mongodb::server::admin_password: password
mongodb::server::keyfile: /var/lib/mongodb/mongodb-keyfile
mongodb::server::key: <shared_key>
mongodb::server::replset: rsmain
mongodb_replset:
  rsmain:
    ensure: present
    members:
      - vm1:27017
      - vm2:27017
      - vm3:27017
    arbiter:
      vm3:27017
mongodb_db:
  my_db:
    user: myuser
    password: my_password
    roles:
      - dbAdmin
[...]
```
vm2 and vm3 (conf sample)
```yaml
[...]
mongodb::server::keyfile: /var/lib/mongodb/mongodb-keyfile
mongodb::server::key: <shared_key>
mongodb::server::replset: rsmain
[...]
```
Step1 => build the arb vm => OK
Step2 => build the db2 vm => OK
Step3 => build the db1 vm => KO



Sample output for vm1
The provider mongodb_replset don't use the auth file.
```shell
Error: Could not prefetch mongodb_replset provider 'mongo': Execution of '/usr/bin/mongo --quiet --host 127.0.0.1:27017 --eval printjson(rs.conf())' returned 252: 2015-05-15T16:21:26.088+0200 error: { "$err" : "not authorized for query on local.system.replset", "code" : 13 } at src/mongo/shell/query.js:131
```
Once the admin user created, the command:
```shell
Execution of '/usr/bin/mongo --quiet --host 127.0.0.1:27017 --eval printjson(rs.conf())'
```
would be something like:
```shell
Execution of '/usr/bin/mongo --quiet --host 127.0.0.1:27017 --eval load('/root/.mongorc.js'); printjson(rs.conf())
```

I'm not sure, but maybe we have to modify the file [lib/puppet/provider/mongodb_replset/mongo.rb](https://github.com/puppetlabs/puppetlabs-mongodb/blob/master/lib/puppet/provider/mongodb_replset/mongo.rb) and add authentication support in replicaset provider.
Something like
```ruby
   18   commands :mongo => 'mongo'                                                    
   19                                                                                 
+  20   # Optional defaults file                                                      
+  21   def self.mongorc_file                                                         
+  22     if File.file?("#{Facter.value(:root_home)}/.mongorc.js")                    
+  23       "load('#{Facter.value(:root_home)}/.mongorc.js'); "                       
+  24     else                                                                        
+  25       nil                                                                       
+  26     end                                                                         
+  27   end                                                                           
+  28                                                                                 
+  29   def mongorc_file                                                              
+  30     self.class.mongorc_file                                                     
+  31   end                                                                           
+  32                 
```
And replace `mongo_command` by `mongo_eval`

```ruby
   82   def db_ismaster(host)                                                         
~  83     mongo_eval("db.isMaster()", host)                                           
   84   end  
```
And finaly, add support for command with the rc file
```ruby
+ 253         if mongorc_file                                                         
+ 254           args << ['--eval',"load('/root/.mongorc.js'); printjson(#{command})"] 
+ 255         else                                                                    
+ 256           args << ['--eval',"printjson(#{command})"]                            
+ 257         end                                                                     
+ 258       #end                                  
```

Last thing, we also need 'root' role for admin user to execute replicaset commands.
With roles created by the Puppet module (userAdmin, dbAdmin, dbOwner, userAdminAnyDatabase)

```shell
vagrant@vm1:~$ sudo mongo --eval "load('/root/.mongorc.js'); printjson(rs.conf())"
MongoDB shell version: 2.6.9
connecting to: test
2015-05-15T16:33:56.462+0200 error: { "$err" : "not authorized for query on local.system.replset", "code" : 13 } at src/mongo/shell/query.js:131
```
With the root role
```shell
vm1:PRIMARY> db.grantRolesToUser("admin",[{ role: "root", db: "admin" },])
```
```shell
vagrant@vm1:~$ sudo mongo --eval "load('/root/.mongorc.js'); printjson(rs.conf())"MongoDB shell version: 2.6.9
connecting to: test
{
	"_id" : "rsmain",
	"version" : 1,
	"members" : [
		{
			"_id" : 0,
			"host" : "vm1:27017"
		}
	]
}
```

I hope it will helps.

Bye

fatmcgav commented:
@zeysh Cheers for the feedback and testing with replsets...

I must admin I'm new to MongoDB, so hadn't got to setting up replicas yet... But your examples should get me up and running... 

I'll look at making the tweaks outlined above soon and get the PR updated... 

@tphoney @hunner hold off on merging this until have sorted the above :) 

Cheers
Gavin

fatmcgav commented:
@zeysh I've just pushed a new commit which should cover off the replset issues you had. Give it a test and let me know how you get on... 

Once all good, I'll rebase and then should be GTM :)

zeysh commented:
@fatmcgav, thanks for the tweaks.

After some tests, the first run include the file mongorc.js and the initialisation of the replicaset doesn't work.

```shell
Error: Could not prefetch mongodb_user provider 'mongodb': Execution of '/usr/bin/mongo admin --quiet --host 127.0.0.1:27017 --eval load('/root/.mongorc.js'); printjson(db.system.users.find().toArray())' returned 252: 2015-05-28T00:53:32.286+0200 error: { "$err" : "not master and slaveOk=false", "code" : 13435 } at src/mongo/shell/query.js:131
```
```shell
Debug: Executing '/usr/bin/mongo --quiet --host 127.0.0.1:27017 --eval load('/root/.mongorc.js') printjson(rs.conf())'
==> vm1: Error: Could not prefetch mongodb_replset provider 'mongo': Execution of '/usr/bin/mongo --quiet --host 127.0.0.1:27017 --eval load('/root/.mongorc.js') printjson(rs.conf())' returned 1: 2015-05-28T01:46:40.985+0200 Error: [printjson(rs.conf())] is not a valid database name at src/mongo/shell/mongo.js:41
==> vm1: exception: connect failed
```
```shell
==> vm1: Debug: Executing '/usr/bin/mongo --quiet --host vm1:27017 --eval load('/root/.mongorc.js') printjson(rs.status())'
==> vm1: Warning: Can't connect to replicaset member vm1:27017.
==> vm1: Debug: Checking replicaset member vm2:27017 ...
==> vm1: Debug: Executing '/usr/bin/mongo --quiet --host vm2:27017 --eval load('/root/.mongorc.js') printjson(rs.status())'
==> vm1: Debug: Checking replicaset member vm3:27017 ...
==> vm1: Warning: Can't connect to replicaset member vm2:27017.
==> vm1: Debug: Executing '/usr/bin/mongo --quiet --host vm3:27017 --eval load('/root/.mongorc.js') printjson(rs.status())'
==> vm1: Debug: Class[Company::Profile::Mongodb_replset_node]: The container Stage[main] will propagate my refresh event
==> vm1: Warning: Can't connect to replicaset member vm3:27017.
==> vm1: Error: /Stage[main]/Company::Profile::Mongodb_replset_node/Mongodb_replset[rsmain]: Could not evaluate: Can't connect to any member of replicaset rsmain.

```
```shell
> rs.status()
{
	"startupStatus" : 3,
	"info" : "run rs.initiate(...) if not yet done for the set",
	"ok" : 0,
	"errmsg" : "can't get local.system.replset config from self or any seed (EMPTYCONFIG)"
}
```
Maybe we have to configure the replica like this:

1. Connect to the primary node without authentication
2. Init the replicaset on the primary node: 
```shell
rs.initialisation()
```
3. Add the secondary node:
```shell
rs.add('vm2:27017')
```
4. Add the arbiter node: 
```shell
rs.addArb('vm3:27017')
```
5. After that the /root/.mongorc.js file must be used for every request.

But, I got the following errors when I run puppet again.
```shell
Error: Could not prefetch mongodb_replset provider 'mongo': Execution of '/usr/bin/mongo --quiet --host 127.0.0.1:27017 --eval load('/root/.mongorc.js') printjson(rs.conf())' returned 1: 2015-05-28T02:10:48.749+0200 Error: [printjson(rs.conf())] is not a valid database name at src/mongo/shell/mongo.js:41
exception: connect failed
```

```shell
Notice: /Stage[main]/Company::Profile::Mongodb_replset_node/Mongodb_replset[rsmain]/ensure: created
Warning: Can't connect to replicaset member vm1:27017.
Warning: Can't connect to replicaset member vm2:27017.
Warning: Can't connect to replicaset member vm3:27017.
Error: /Stage[main]/Company::Profile::Mongodb_replset_node/Mongodb_replset[rsmain]: Could not evaluate: Can't connect to any member of replicaset rsmain.
```

For vm1 and vm2, connection failed. But for vm3 (the arbiter) I think it's normal.
Manualy, the connection works...
```shell
vagrant@vm1:~$ mongo -u admin -p password vm1:27017 --authenticationDatabase admin
MongoDB shell version: 2.6.9
connecting to: vm1:27017/test
rsmain:PRIMARY> 
bye
vagrant@vm1:~$ mongo -u admin -p password vm2:27017 --authenticationDatabase admin
MongoDB shell version: 2.6.9
connecting to: vm2:27017/test
mpa_ihm:SECONDARY> ^C
bye
vagrant@vm1:~$ mongo -u admin -p password vm3:27017 --authenticationDatabase admin
MongoDB shell version: 2.6.9
connecting to: vm3:27017/test
2015-05-28T02:22:32.045+0200 Error: 18 { ok: 0.0, errmsg: "auth failed", code: 18 } at src/mongo/shell/db.js:1260
exception: login failed
```

... but failed with the file .mongorc.js.
```
sudo mongo --eval "load('/root/.mongorc.js') printjson(rs.conf())"
MongoDB shell version: 2.6.9
connecting to: test
2015-05-28T02:27:35.657+0200 SyntaxError: Unexpected identifier
```

fatmcgav commented:
@zeysh Cheers for the feedback... 

Could you gist ure puppet test manifests, and I'll see if I can replicate/fix here... 

Gav

zeysh commented:
@fatmcgav 

I just created a vagrant repo for testing: https://github.com/zeysh/vagrant-mongo

Enjoy

fatmcgav commented:
@zeysh Cheers for that, I'll give it a whirl...

fatmcgav commented:
@zeysh Think I've got a bit further by fixing a bug in the mongo_replset provider which was causing the command to fail with 'is not a valid database name'... 

Still need to look at the missing initialise step...

fatmcgav commented:
@zeysh Are you able to provide confirm that automatic replica set setup worked without adding auth to the mix?

As I've tried stripping back to before I added auth support, and I still cant get it to setup a replica set without any manual interaction... 

Cheers
Gav

mhaskel commented:
@fatmcgav ping?

fatmcgav commented:
@mhaskel pong... 😉

I need to come back to this... Will try to update early next week... 

Gav

fatmcgav commented:
Am in

fatmcgav commented:
Hmm, that rebase got messy... Will try and tidy up once I've got something working...

mhaskel commented:
Great, thanks for taking a look @fatmcgav !

fatmcgav commented:
OK, I think I've cracked it... 

Turns out that MongoDB, atleast at 2.6.9, returns an invalid JSON response when attempting to connect to a remote DB with auth enabled but replicaset not yet setup... 
E.g.:
```ruby
==> primary: Debug: Exception raised when attempting to parse JSON output: 757: unexpected token at 'Error: 18 { ok: 0.0, errmsg: "auth failed", code: 18 }
==> primary: {
==> primary:    "ok" : 0,
==> primary:    "errmsg" : "not authorized on admin to execute command { replSetGetStatus: 1.0 }",
==> primary:    "code" : 13
==> primary: }
==> primary: '
```

So I put in a bit of a hack to rescue a JSON.ParserError, strip off the first line and then parse it again... 
Then with a couple of tweaks in to handle the unauthorized response, it all kicked into life... 

Have pushed the latest code to my [support_auth_fix](https://github.com/fatmcgav/puppetlabs-mongodb/tree/support_auth_fix) branch. 

@zeysh Are you able to give it another go and see what you can do?
I've also updated the vagrant-mongo env you setup - latest is [here](https://github.com/fatmcgav/vagrant-mongo)

Let me know... 

In the mean time, I'll run through the acceptance tests and see how they behave... 

Cheers
Gav

fatmcgav commented:
OK, I've rebased and squashed this branch off of my support_auth_fix branch, so this should now be good to merge. 

Am as confident as I can be that this should work :) 

Cheers
Gav

fatmcgav commented:
Scratch that, spotted a couple more issues today with trying to use the module in anger.

Should have rest of the stuff wrapped up tomorrow...

Gav

mhaskel commented:
Great, thanks @fatmcgav ! Also, unfortunately looks like another merge has this needing a rebase again, so if you could look at that that would be great (and sorry for the inconvenience!)

fatmcgav commented:
@mhaskel Have rebased and updated. Had to re-add the stuff pulled out by 31ce292f4255b3b115d64f94bb74703a10210a14 as I couldn't get it working without... 

So not sure if anyone who knows MongoDB can review https://github.com/puppetlabs/puppetlabs-mongodb/pull/194#commitcomment-13065305

Cheers
Gav

mhaskel commented:
@nibalizer do you know anything about mongodb (or enough to review this)? Or, know who might?

fatmcgav commented:
@iberezovskiy has reached out to me with a view to integrating PR #231 with this one, as it fixes a couple of issues that I've spotted but yet to get fixed, especially the 'slave permissions' issue. 

I'm going to work on getting integrating into a new branch that we can both test on, and then when we're happy will get this PR updated and hopefully merged in...

fatmcgav commented:
@mhaskel OK, I think that latest push should have it licked... 

Thanks to @iberezovskiy for the assist and testing the code aswell. 

Let me know if anything is missing for a merge :) 

Cheers
Gav

iberezovskiy commented:
This patch now looks very good for me.
First of all, this patch improves replica set initialization when auth is enabled.
Second, commands for users/databases can be run only from master node now (it prevent users from errors during puppet run on slave nodes)
Third, ability to create admin was added, and it runs only after replica set initialization (if replica is enabled, of course).

Thanks Gavin for hard work on this big patch!

timogoebel commented:
@fatmcgav: Thanks for the great work. I have one further suggestion though:
Right now you have to set either $replset_members or $replica_sets. I think, this should be optional if I want to take care of this somewhere else, e.g. another manifest. What do you think?

fatmcgav commented:
@timogoebel Technically there's nothing stopping both params being optional. 

The biggest challenge is ensuring the resource ordering is correct. 

Though with the changes to mongodb::replset class in *should* work... As long as you're using the mongodb::replset class to configure it. 

I'll run some tests see if it's feasible.. 

Cheers 
Gavin

fatmcgav commented:
@timogoebel I've just added 86af881 which makes the replicaset config in mongodb::server optional. 

It does throw a warning, and the assumption is that you're using mongodb::replset in order to get the ordering correct. I've tested using [vagrant-mongo](https://github.com/fatmcgav/vagrant-mongo) and it seems to work OK... 

Let me know your thoughts... 

@iberezovskiy I went and renamed the `replica_sets` param to `replset_config`, as I think it fits better with the other params and what it's actually used for... LMK your thoughts... 

@mhaskel Will get this rebased once the latest commits have been reviewed. Hopefully can then get it merged in :) 

Cheers
Gav

timogoebel commented:
@fatmcgav : Looks fine to me, thanks a lot for your awesome work! I already backported some of your changes and they are working fine in our environment! Hope, this gets merged soon!

fatmcgav commented:
@timogoebel Cheers for the confirmation...

timogoebel commented on lib/puppet/provider/mongodb_replset/mongo.rb:
```diff
@@ -201,12 +232,33 @@ def set_members
       # Add members to an existing replset
+      Puppet.debug "Adding member to existing replset #{self.name}"
       if master = master_host(alive_hosts)
         current_hosts = db_ismaster(master)['hosts']
```
@fatmcgav: This should get arbiters as well or else it will try to add the arbiter even though it is already added.

I will try to prepare a patch for this.

```
services-lxstage:PRIMARY> db.isMaster()
{
	"setName" : "services-XXX",
	"setVersion" : 1,
	"ismaster" : true,
	"secondary" : false,
	"hosts" : [
		"services-mongodb-02.XX:27017"
	],
	"arbiters" : [
		"services-mongodb-00.XX:27017"
	],
	"primary" : "services-mongodb-02.XX:27017",
	"me" : "services-mongodb-02.XX:27017",
	"electionId" : ObjectId("5604f2a2c3f732daff9b4d6c"),
	"maxBsonObjectSize" : 16777216,
	"maxMessageSizeBytes" : 48000000,
	"maxWriteBatchSize" : 1000,
	"localTime" : ISODate("2015-09-25T08:07:01.145Z"),
	"maxWireVersion" : 2,
	"minWireVersion" : 0,
	"ok" : 1
}
```

timogoebel replied:
This fixes this issue:

```
-        current_hosts = db_ismaster(master)['hosts']
+        master_data = db_ismaster(master)
+        current_hosts = master_data['hosts']
+        current_hosts = current_hosts + master_data['arbiters'] if master_data.has_key?('arbiters')
```

fatmcgav replied:
@timogoebel Sounds sensible... However I've not been able to replicate this repeat addition... 

Are you able to gist a debug log? 

Cheers

fatmcgav replied:
Ah, managed to replicate it :) 

Will get those changes added in :) 

Cheers

iberezovskiy commented:
@fatmcgav : +1 for this patch. It works in my environment too. Thanks again for awesome work!

fatmcgav commented:
@iberezovskiy Cheers for that... 

@mhaskel I've squashed the commits down so think this is now looking good to merge... 

LMK if you need anything further...

fatmcgav commented:
@jsosic OK, think I managed to come up with a fix to remove the use of `mongo` in the `mongodb_replset` provider. 

All the existing tests pass locally, so will get this squashed later if no-one objects... 

@timogoebel @iberezovskiy  A quick sanity tests would be appreciated. 

Cheers
Gavin

iberezovskiy commented:
@fatmcgav It works for my deployment well. LGTM, +1.

fatmcgav commented:
Cheers for the confirm @iberezovskiy 

I've squashed the commits down and updated the branch, so should be good to merge soon :) 

I am looking at adding some facts to support this functionality, but I'll probably raise those on a separate PR, as think this one has been going long enough already :)

squarebracket commented:
@fatmcgav there's a typo on line 411 of `README.md` -- it says `replet` instead of `replset`

fatmcgav commented:
Good spot... Will get that fixed up...

squarebracket commented:
@fatmcgav My pleasure, thanks for the code!

fatmcgav commented:
OK, that's the latest tweaks all squashed down... 

@mhaskel I think we're ready for merge now... Unless I've missed something...

mhaskel commented:
@hunner @bmjen @tphoney want to take a look?

mentat commented:
It looks like there are quite a few acceptance test failures here.  Could you take a look at those?

fatmcgav commented:
@mentat I thought they all passed last time I ran them locally... Can take a look again though. 

Are you looking at a CI env, or running locally? 

Cheers

timogoebel commented:
@fatmcgav : I hope, you just closed this by accident.

fatmcgav commented:
@timogoebel I did... Reopened... :-)

mentat commented:
@fatmcgav I ran the acceptance tests locally.

fatmcgav commented:
@mentat OK, I've managed to get it down to 2 failing tests... 

The failures now are related to non-idempotent Password hash generation, which I'm not sure is easily work-around'able... Happy for any suggestions... 

```
Failures:

  1) mongodb_database behaves like normal tests when manage_package_repo is false and version is  creating a database with default port should compile with no errors
     Failure/Error: apply_manifest(pp, :catch_changes  => true)
     Beaker::Host::CommandFailure:
       Host 'centos-65-x64' exited with 2 running:
        puppet apply --verbose --detailed-exitcodes /tmp/apply_manifest.pp.7uhdRT
       Last 10 lines of output were:
        Info: Loading facts
        Info: Loading facts
        Notice: Compiled catalog for centos-65-x64.localdomain in environment production in 1.47 seconds
        Info: Applying configuration version '1445612057'
        Notice: /Stage[main]/Main/Mongodb::Db[testdb1]/Mongodb_user[User testuser on db testdb1]/password_hash: password_hash changed '84b719f4c9000b640c3f1bfe3257341e' to '19c8747ac7b3654045c99bea032d61be'
        Notice: /Stage[main]/Main/Mongodb::Db[testdb2]/Mongodb_user[User testuser on db testdb2]/password_hash: password_hash changed '84b719f4c9000b640c3f1bfe3257341e' to '19c8747ac7b3654045c99bea032d61be'
        Notice: Finished catalog run in 0.69 seconds

     Shared Example Group: "normal tests" called from ./spec/acceptance/database_spec.rb:96
     # /Users/gavinw/.rvm/gems/ruby-1.9.3-p551@puppetlabs-mongodb/gems/beaker-2.11.0/lib/beaker/host.rb:300:in `exec'
     # /Users/gavinw/.rvm/gems/ruby-1.9.3-p551@puppetlabs-mongodb/gems/beaker-2.11.0/lib/beaker/dsl/helpers/host_helpers.rb:73:in `block in on'
     # /Users/gavinw/.rvm/gems/ruby-1.9.3-p551@puppetlabs-mongodb/gems/beaker-2.11.0/lib/beaker/shared/host_manager.rb:104:in `run_block_on'
     # /Users/gavinw/.rvm/gems/ruby-1.9.3-p551@puppetlabs-mongodb/gems/beaker-2.11.0/lib/beaker/dsl/patterns.rb:32:in `block_on'
     # /Users/gavinw/.rvm/gems/ruby-1.9.3-p551@puppetlabs-mongodb/gems/beaker-2.11.0/lib/beaker/dsl/helpers/host_helpers.rb:63:in `on'
     # /Users/gavinw/.rvm/gems/ruby-1.9.3-p551@puppetlabs-mongodb/gems/beaker-2.11.0/lib/beaker/dsl/helpers/puppet_helpers.rb:435:in `block in apply_manifest_on'
     # /Users/gavinw/.rvm/gems/ruby-1.9.3-p551@puppetlabs-mongodb/gems/beaker-2.11.0/lib/beaker/shared/host_manager.rb:104:in `run_block_on'
     # /Users/gavinw/.rvm/gems/ruby-1.9.3-p551@puppetlabs-mongodb/gems/beaker-2.11.0/lib/beaker/dsl/patterns.rb:32:in `block_on'
     # /Users/gavinw/.rvm/gems/ruby-1.9.3-p551@puppetlabs-mongodb/gems/beaker-2.11.0/lib/beaker/dsl/helpers/puppet_helpers.rb:364:in `apply_manifest_on'
     # /Users/gavinw/.rvm/gems/ruby-1.9.3-p551@puppetlabs-mongodb/gems/beaker-2.11.0/lib/beaker/dsl/helpers/puppet_helpers.rb:442:in `apply_manifest'
     # ./spec/acceptance/database_spec.rb:44:in `block (6 levels) in <top (required)>'

  2) mongodb_database behaves like normal tests when manage_package_repo is false and version is  creating a database with custom port should work with no errors
     Failure/Error: apply_manifest(pp, :catch_changes  => true)
     Beaker::Host::CommandFailure:
       Host 'centos-65-x64' exited with 2 running:
        puppet apply --verbose --detailed-exitcodes /tmp/apply_manifest.pp.V7SgPp
       Last 10 lines of output were:
        Info: Loading facts
        Info: Loading facts
        Notice: Compiled catalog for centos-65-x64.localdomain in environment production in 1.53 seconds
        Info: Applying configuration version '1445612109'
        Notice: /Stage[main]/Main/Mongodb::Db[testdb1]/Mongodb_user[User testuser on db testdb1]/password_hash: password_hash changed '84b719f4c9000b640c3f1bfe3257341e' to '19c8747ac7b3654045c99bea032d61be'
        Notice: /Stage[main]/Main/Mongodb::Db[testdb2]/Mongodb_user[User testuser on db testdb2]/password_hash: password_hash changed '84b719f4c9000b640c3f1bfe3257341e' to '19c8747ac7b3654045c99bea032d61be'
        Notice: Finished catalog run in 0.83 seconds

     Shared Example Group: "normal tests" called from ./spec/acceptance/database_spec.rb:96
     # /Users/gavinw/.rvm/gems/ruby-1.9.3-p551@puppetlabs-mongodb/gems/beaker-2.11.0/lib/beaker/host.rb:300:in `exec'
     # /Users/gavinw/.rvm/gems/ruby-1.9.3-p551@puppetlabs-mongodb/gems/beaker-2.11.0/lib/beaker/dsl/helpers/host_helpers.rb:73:in `block in on'
     # /Users/gavinw/.rvm/gems/ruby-1.9.3-p551@puppetlabs-mongodb/gems/beaker-2.11.0/lib/beaker/shared/host_manager.rb:104:in `run_block_on'
     # /Users/gavinw/.rvm/gems/ruby-1.9.3-p551@puppetlabs-mongodb/gems/beaker-2.11.0/lib/beaker/dsl/patterns.rb:32:in `block_on'
     # /Users/gavinw/.rvm/gems/ruby-1.9.3-p551@puppetlabs-mongodb/gems/beaker-2.11.0/lib/beaker/dsl/helpers/host_helpers.rb:63:in `on'
     # /Users/gavinw/.rvm/gems/ruby-1.9.3-p551@puppetlabs-mongodb/gems/beaker-2.11.0/lib/beaker/dsl/helpers/puppet_helpers.rb:435:in `block in apply_manifest_on'
     # /Users/gavinw/.rvm/gems/ruby-1.9.3-p551@puppetlabs-mongodb/gems/beaker-2.11.0/lib/beaker/shared/host_manager.rb:104:in `run_block_on'
     # /Users/gavinw/.rvm/gems/ruby-1.9.3-p551@puppetlabs-mongodb/gems/beaker-2.11.0/lib/beaker/dsl/patterns.rb:32:in `block_on'
     # /Users/gavinw/.rvm/gems/ruby-1.9.3-p551@puppetlabs-mongodb/gems/beaker-2.11.0/lib/beaker/dsl/helpers/puppet_helpers.rb:364:in `apply_manifest_on'
     # /Users/gavinw/.rvm/gems/ruby-1.9.3-p551@puppetlabs-mongodb/gems/beaker-2.11.0/lib/beaker/dsl/helpers/puppet_helpers.rb:442:in `apply_manifest'
     # ./spec/acceptance/database_spec.rb:85:in `block (6 levels) in <top (required)>'

Finished in 6 minutes 3 seconds (files took 2 minutes 32.2 seconds to load)
12 examples, 2 failures

Failed examples:

rspec ./spec/acceptance/database_spec.rb:28 # mongodb_database behaves like normal tests when manage_package_repo is false and version is  creating a database with default port should compile with no errors
rspec ./spec/acceptance/database_spec.rb:69 # mongodb_database behaves like normal tests when manage_package_repo is false and version is  creating a database with custom port should work with no errors
/Users/gavinw/.rvm/rubies/ruby-1.9.3-p551/bin/ruby -I/Users/gavinw/.rvm/gems/ruby-1.9.3-p551@puppetlabs-mongodb/gems/rspec-core-3.1.7/lib:/Users/gavinw/.rvm/gems/ruby-1.9.3-p551@puppetlabs-mongodb/gems/rspec-support-3.1.2/lib /Users/gavinw/.rvm/gems/ruby-1.9.3-p551@puppetlabs-mongodb/gems/rspec-core-3.1.7/exe/rspec spec/acceptance/database_spec.rb --color failed
```

mentat commented:
@fatmcgav great work.  Those tests are failing in master too, so I would not worry about fixing them for this PR.

fatmcgav commented:
@mentat Cool, cheers for confirming. Will wait for green on my latest commit, then I'll squash down...

fatmcgav commented:
It's GREEN!!! Have squashed and force updated :) 

Hopeful GTM now :)

fatmcgav commented:
WOOH!!! :) Cheers for that @mentat
